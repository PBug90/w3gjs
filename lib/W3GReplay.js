const fs = require('fs')
const zlib = require('zlib')
const { Parser } = require('binary-parser')
const { ActionBlockList } = require('../parsers/actions')
const { ReplayHeader,
  EncodedMapMetaString,
  GameMetaData} = require('../parsers/header')

const { GameDataParser } = require('../parsers/gamedata')

const GameDataParserComposed = new Parser()
  .nest('meta', {type: GameMetaData})
  .nest('blocks', {type: GameDataParser})

const Player = require('./Player')
const crypto = require('crypto')

function W3GReplay () {
  return this
}

W3GReplay.prototype.parse = function (filepath) {
  console.time('parse')
  this.buffer = fs.readFileSync(filepath)
  this.buffer = this.buffer.slice(this.buffer.indexOf('Warcraft III recorded game'))

  this.header = ReplayHeader.parse(this.buffer)
  const decompressed = []
  this.header.blocks.forEach((block) => {
    if (block.blockSize > 0 && block.blockDecompressedSize === 8192) {
      try {
        const r = zlib.inflateSync(block.compressed, { finishFlush: zlib.constants.Z_SYNC_FLUSH })
        if (r.byteLength > 0 && block.compressed.byteLength > 0) {
          decompressed.push(r)
        }
      } catch (ex) {
        console.log(ex)
      }
    }
  })
  this.decompressed = Buffer.concat(decompressed)
  this.gameMetaDataDecoded = GameDataParserComposed.parse(this.decompressed)

  this.decodedMetaStringBuffer = this.decodeGameMetaString(this.gameMetaDataDecoded.meta.encodedString)
  this.meta = Object.assign({}, this.gameMetaDataDecoded, EncodedMapMetaString.parse(this.decodedMetaStringBuffer))

  this.slots = this.gameMetaDataDecoded.meta.playerSlotRecords
  this.playerList = [this.gameMetaDataDecoded.meta.player, ...this.gameMetaDataDecoded.meta.playerList]

  this.playerActionTrackInterval = 60000 // apm tracking interval in milliseconds

  this.createPlayerList()
  this.processGameDataBlocks()
  this.cleanup()

  return this.finalize()
}

W3GReplay.prototype.createPlayerList = function () {
  const tempPlayers = {}
  this.teams = {}
  this.players = {}

  this.playerList.forEach(player => {
    tempPlayers[player.playerId] = player
  })

  this.slots.forEach(slot => {
    if (slot.slotStatus > 1) {
      this.players[slot.playerId] = { ...this.players[slot.playerId], ...slot }
      this.teams[slot.teamId] = this.teams[slot.teamId] || []
      this.teams[slot.teamId].push(slot.playerId)
      this.players[slot.playerId] =
        new Player(slot.playerId, tempPlayers[slot.playerId] ? tempPlayers[slot.playerId].playerName : 'Computer', slot.teamId, slot.color, slot.raceFlag)
    }
  })
}

W3GReplay.prototype.processGameDataBlocks = function () {
  this.chatlog = []
  this.leaveEvents = []
  this.w3mmd = []

  this.totalTimeTracker = 0
  this.timeSegmentTracker = 0
  this.gameMetaDataDecoded.blocks.forEach((block) => {
    switch (block.type) {
      case 31:
      case 30:
        this.totalTimeTracker += block.timeIncrement
        this.timeSegmentTracker += block.timeIncrement
        if (this.timeSegmentTracker > this.playerActionTrackInterval) {
          Object.values(this.players).forEach((p) => p.newActionTrackingSegment())
          this.timeSegmentTracker = 0
        }
        this.processTimeSlot(block)
        break
      case 32:
        block.timeMS = this.totalTimeTracker
        this.chatlog.push(block)
        break
      case 23:
        this.leaveEvents.push(block)
        break
    }
  })

  this.chatlog = this.chatlog.map(elem => ({...elem, playerName: this.players[elem.playerId].name}))
  delete this.temporaryAPMTracker
}

W3GReplay.prototype.processTimeSlot = function (timeSlotBlock) {
  timeSlotBlock.actions.forEach(actionBlock => {
    this.processCommandDataBlock(actionBlock)
  })
}

W3GReplay.prototype.processCommandDataBlock = function (actionBlock) {
  const currentPlayer = this.players[actionBlock.playerId]
  let lastActionWasDeselect = false
  currentPlayer.currentTimePlayed = this.totalTimeTracker
  try {
    ActionBlockList.parse(actionBlock.actions).forEach(action => {
      switch (action.actionId) {
        case 0x10:
          currentPlayer.handle0x10(action.itemId, this.totalTimeTracker)
          break
        case 0x11:
          currentPlayer.handle0x11(action.itemId, this.totalTimeTracker)
          break
        case 0x12:
          currentPlayer.handle0x12(action.itemId, this.totalTimeTracker)
          break
        case 0x13:
          currentPlayer.handle0x13(action.itemId)
          break
        case 0x14:
          currentPlayer.handle0x14(action.itemId1)
          break
        case 0x16:
          if (action.selectMode === 0x02) {
            lastActionWasDeselect = true
            currentPlayer.handle0x16(action.selectMode, true)
          } else {
            if (lastActionWasDeselect === false) {
              currentPlayer.handle0x16(action.selectMode, true)
            }
            lastActionWasDeselect = false
          }
          break
        case 0x17:
        case 0x18:
        case 0x1C:
        case 0x1D:
        case 0x1E:
        case 0x61:
        case 0x65:
        case 0x66:
        case 0x67:
          currentPlayer.handleOther(action.actionId)
          break
        case 0x6b:
          this.w3mmd.push(action)
          break
      }
    })
  } catch (ex) {
    console.error(ex)
  }
}

W3GReplay.prototype.decodeGameMetaString = function (str) {
  let test = Buffer.from(str, 'hex')
  let decoded = Buffer.alloc(test.length)
  let mask = 0
  let dpos = 0

  for (let i = 0; i < test.length; i++) {
    if (i % 8 === 0) {
      mask = test[i]
    } else {
      if ((mask & (0x1 << (i % 8))) === 0) {
        decoded.writeUInt8(test[i] - 1, dpos++)
      } else {
        decoded.writeUInt8(test[i], dpos++)
      }
    }
  }
  return decoded
}

W3GReplay.prototype.isObserver = function (player) {
  return (player.teamid === 24 && this.header.version >= 29) || (player.teamid === 12 && this.header.version < 29)
}

W3GReplay.prototype.determineMatchup = function () {
  let teamRaces = {}
  Object.values(this.players).forEach((p) => {
    if (!this.isObserver(p)) {
      teamRaces[p.teamid] = teamRaces[p.teamid] || []
      teamRaces[p.teamid].push(p.raceDetected || p.race)
    }
  })
  this.gametype = Object.values(teamRaces).map(e => e.length).sort().join('on')
  this.matchup = Object.values(teamRaces).map(e => e.sort().join('')).sort().join('v')
}

W3GReplay.prototype.generateID = function () {
  let players = Object.values(this.players).filter((p) => this.isObserver(p) === false).sort((player1, player2) => {
    if (player1.id < player2.id) {
      return -1
    }
    return 1
  }).reduce((accumulator, player) => {
    accumulator += `${player.name}_${player.id}_${player.teamid}`
    return accumulator
  }, '')
  const idBase = this.meta.meta.randomSeed + players + this.meta.mapName
  this.id = crypto.createHash('sha256').update(idBase).digest('hex')
}

W3GReplay.prototype.cleanup = function () {
  this.determineMatchup()
  this.generateID()
  this.observers = []

  Object.values(this.players).forEach(p => {
    p.newActionTrackingSegment(this.playerActionTrackInterval)
    p.cleanup()
    if (this.isObserver(p)) {
      this.observers.push(p.name)
      delete this.players[p.id]
    }
  })

  if (this.header.version >= 29 && this.teams.hasOwnProperty('24')) {
    delete this.teams['24']
  } else if (this.teams.hasOwnProperty('12')) {
    delete this.teams['12']
  }
  delete this.slots
  delete this.playerList
  delete this.buffer
  delete this.decompressed
  delete this.gameMetaDataDecoded
  delete this.decodedMetaStringBuffer
  delete this.header.blocks
  delete this.apmTimeSeries
  delete this.meta.blocks
}

W3GReplay.prototype.finalize = function () {
  const settings = {
    referees: Boolean(this.meta.referees),
    fixedTeams: Boolean(this.meta.fixedTeams),
    fullSharedUnitControl: Boolean(this.meta.fullSharedUnitControl),
    alwaysVisible: Boolean(this.meta.alwaysVisible),
    hideTerrain: Boolean(this.meta.hideTerrain),
    mapExplored: Boolean(this.meta.mapExplored),
    teamsTogether: Boolean(this.meta.teamsTogether),
    randomHero: Boolean(this.meta.randomHero),
    randomRaces: Boolean(this.meta.randomRaces),
    speed: this.meta.speed
  }
  const root = {
    id: this.id,
    gamename: this.meta.meta.gameName,
    randomseed: this.meta.meta.randomSeed,
    startSpots: this.meta.meta.startSpotCount,
    observers: this.observers,
    players: Object.values(this.players).sort((player1, player2) => player2.teamid >= player1.teamid && player2.id > player1.id ? -1 : 1),
    matchup: this.matchup,
    creator: this.meta.creator,
    type: this.gametype,
    chat: [],
    apm: {
      trackingInterval: this.playerActionTrackInterval
    },
    map: {
      path: this.meta.mapName,
      file: this.meta.mapName.split('\\').pop(),
      checksum: this.meta.mapChecksum
    },
    version: `1.${this.header.version}`,
    duration: this.header.replayLengthMS,
    expansion: this.header.gameIdentifier === 'PX3W',
    settings
  }

  return root
}

module.exports = W3GReplay
