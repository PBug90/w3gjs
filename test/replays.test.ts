import W3GReplay from '../src/W3GReplay'
import { Validator } from 'jsonschema'
import { readFileSync } from 'fs'

const Parser = new W3GReplay()

describe('Replay parsing tests', () => {
    it('parses a standard 1.29 replay with observers properly', () => {
        const test = Parser.parse('./replays/standard_129_obs.w3g')

        expect(test.version).toBe('1.29')
        expect(test.players[1].name).toBe('S.o.K.o.L')
        expect(test.players[1].raceDetected).toBe('O')
        expect(test.players[1].id).toBe(4)
        expect(test.players[1].teamid).toBe(3)
        expect(test.players[1].color).toBe('#00781e')
        expect(test.players[1].units.summary).toEqual({
            opeo: 10, ogru: 5, ostr: 1, orai: 6, ospm: 5, okod: 2
        })
        expect(test.players[1].actions).toEqual({
            assigngroup: 38,
            rightclick: 1104,
            basic: 122,
            buildtrain: 111,
            ability: 59,
            item: 6,
            select: 538,
            removeunit: 0,
            subgroup: 0,
            selecthotkey: 751,
            esc: 0,
            timed: expect.any(Array)
        })

        expect(test.players[0].name).toBe('Stormhoof')
        expect(test.players[0].raceDetected).toBe('O')
        expect(test.players[0].color).toBe('#9b0000')
        expect(test.players[0].id).toBe(6)
        expect(test.players[0].teamid).toBe(0)
        expect(test.players[0].units.summary).toEqual({
            opeo: 11, ogru: 8, ostr: 2, orai: 8, ospm: 4, okod: 3
        })
        expect(test.players[0].actions).toEqual({
            assigngroup: 111,
            rightclick: 1595,
            basic: 201,
            buildtrain: 112,
            ability: 57,
            item: 5,
            select: 653,
            removeunit: 0,
            subgroup: 0,
            selecthotkey: 1865,
            esc: 4,
            timed: expect.any(Array)
        })

        expect(test.observers.length).toBe(4)
        expect(test.chat.length).toBeGreaterThan(2)
        expect(test.matchup).toBe('OvO')
        expect(test.type).toBe('1on1')
        expect(test.players.length).toBe(2)
        expect(test.parseTime).toBe(Math.round(test.parseTime))
        expect(test.map).toEqual({
            checksum: '008ab7f1',
            file: 'w3arena__twistedmeadows__v3.w3x',
            path: 'Maps\\w3arena\\w3arena__twistedmeadows__v3.w3x'
        })
    })

    it('parses a standard 1.26 replay properly', () => {
        const test = Parser.parse('./replays/standard_126.w3g')
        expect(test.version).toBe('1.26')
        expect(test.observers.length).toBe(8)
        expect(test.players[1].name).toBe('Happy_')
        expect(test.players[1].raceDetected).toBe('U')
        expect(test.players[1].color).toBe('#0042ff')
        expect(test.players[0].name).toBe('u2.sok')
        expect(test.players[0].raceDetected).toBe('H')
        expect(test.players[0].color).toBe('#ff0303')
        expect(test.matchup).toBe('HvU')
        expect(test.type).toBe('1on1')
        expect(test.players.length).toBe(2)
        expect(test.map).toEqual({
            checksum: '51a1c63b',
            file: 'w3arena__amazonia__v3.w3x',
            path: 'Maps\\w3arena\\w3arena__amazonia__v3.w3x'
        })
    })

    it('parses a netease 1.29 replay properly', () => {
        const test = Parser.parse('./replays/netease_129_obs.nwg')
        expect(test.version).toBe('1.29')

        expect(test.players[1].name).toBe('rudan')
        expect(test.players[1].color).toBe('#282828')
        expect(test.observers.length).toBe(1)
        expect(test.matchup).toBe('NvN')
        expect(test.type).toBe('1on1')
        expect(test.players.length).toBe(2)
        expect(test.map).toEqual({
            checksum: '281f9d6a',
            file: '(4)TurtleRock.w3x',
            path: 'Maps/1.29\\(4)TurtleRock.w3x'
        })
    })

    it('parses a 2on2standard 1.29 replay properly', () => {
        const test = Parser.parse('./replays/999.w3g')
        expect(test.version).toBe('1.26')
        expect(test.players[0].id).toBe(2)
        expect(test.players[0].teamid).toBe(0)
        expect(test.players[1].id).toBe(4)
        expect(test.players[1].teamid).toBe(0)
        expect(test.players[2].id).toBe(3)
        expect(test.players[2].teamid).toBe(1)
        expect(test.players[3].id).toBe(5)
        expect(test.players[3].teamid).toBe(1)
        expect(test.matchup).toBe('HUvHU')
        expect(test.type).toBe('2on2')
        expect(test.players.length).toBe(4)
        expect(test.map).toEqual({
            checksum: 'b4230d1e',
            file: 'w3arena__maelstrom__v2.w3x',
            path: 'Maps\\w3arena\\w3arena__maelstrom__v2.w3x'
        })
    })

    it('parses a standard 1.30 replay properly', () => {
        const test = Parser.parse('./replays/standard_130.w3g')
        expect(test.version).toBe('1.30')
        expect(test.matchup).toBe('NvU')
        expect(test.type).toBe('1on1')
        expect(test.players[0].name).toBe('sheik')
        expect(test.players[0].race).toBe('U')
        expect(test.players[0].raceDetected).toBe('U')
        expect(test.players[1].name).toBe('123456789012345')
        expect(test.players[1].race).toBe('N')
        expect(test.players[1].raceDetected).toBe('N')
        expect(test.players.length).toBe(2)
        expect(test.players[0].heroes[0]).toEqual(expect.objectContaining({ id: 'Udea', level: 6 }))
        expect(test.players[0].heroes[1]).toEqual(expect.objectContaining({ id: 'Ulic', level: 6 }))
        expect(test.players[0].heroes[2]).toEqual(expect.objectContaining({ id: 'Udre', level: 3 }))
        expect(test.map).toEqual({
            file: '(4)TwistedMeadows.w3x',
            checksum: 'c3cae01d',
            path: 'Maps\\FrozenThrone\\(4)TwistedMeadows.w3x'
        })
    })

    it('parsing result has the correct schema', () => {
        const schema = require('./schema.json')
        const test = Parser.parse('./replays/standard_130.w3g')
        const validatorInstance = new Validator()
        validatorInstance.validate(test, schema, { throwError: true })
    })

    it('resets elapsedMS instance property to 0 before parsing another replay', () => {
        Parser.parse('./replays/standard_130.w3g')
        const msElapsed = Parser.msElapsed
        Parser.parse('./replays/standard_130.w3g')
        const msElapsedTwo = Parser.msElapsed
        expect(msElapsed).toEqual(msElapsedTwo)
    })

    it('parses a standard 1.30.2 replay properly', () => {
        const test = Parser.parse('./replays/standard_1302.w3g')
        expect(test.version).toBe('1.30.2+')
        expect(test.matchup).toBe('NvU')
        expect(test.players.length).toBe(2)
    })

    it('parses a standard 1.30.3 replay properly', () => {
        const test = Parser.parse('./replays/standard_1303.w3g')
        expect(test.version).toBe('1.30.2+')
        expect(test.players.length).toBe(2)
    })

    it('parses a standard 1.30.4 replay properly', () => {
        const test = Parser.parse('./replays/standard_1304.w3g')
        expect(test.version).toBe('1.30.2+')
        expect(test.players.length).toBe(2)
    })

    it('parses a standard 1.30.4 replay properly as buffer', () => {
        const buffer: Buffer = readFileSync('./replays/standard_1304.w3g')
        const test = Parser.parse(buffer)
        expect(test.version).toBe('1.30.2+')
        expect(test.players.length).toBe(2)
    })

    it('parses a standard 1.30.4 2on2 replay properly', () => {
        const test = Parser.parse('./replays/standard_1304.2on2.w3g')
        expect(test.version).toBe('1.30.2+')
        expect(test.players.length).toBe(4)
    })

    it('parses a standard 1.30.4 1on1 tome of retraining', () => {
        const test = Parser.parse('./replays/standard_tomeofretraining_1.w3g')
        expect(test.version).toBe('1.31')
        expect(test.players.length).toBe(2)
        expect(test.players[0].heroes[0]).toEqual({
            id: 'Hamg',
            abilities: {
                AHab: 2,
                AHbz: 2
            },
            retrainingHistory: [
                {
                    abilities: {
                        AHab: 2,
                        AHwe: 2
                    },
                    time: 1136022
                }
            ],
            level: 4,
            abilityOrder: [
                {
                    time: 124366,
                    type: 'ability',
                    value: 'AHwe'
                },
                {
                    time: 234428,
                    type: 'ability',
                    value: 'AHab'
                },
                {
                    time: 293007,
                    type: 'ability',
                    value: 'AHwe'
                },
                {
                    time: 1060007,
                    type: 'ability',
                    value: 'AHab'
                },
                {
                    time: 1136022,
                    type: 'retraining'
                },
                {
                    time: 1140944,
                    type: 'ability',
                    value: 'AHbz'
                },
                {
                    time: 1141147,
                    type: 'ability',
                    value: 'AHbz'
                },
                {
                    time: 1141460,
                    type: 'ability',
                    value: 'AHab'
                },
                {
                    time: 1141569,
                    type: 'ability',
                    value: 'AHab'
                }
            ]
        })
    })

    it('parses a replay with action 0x7a successfully', () => {
        const test = Parser.parse('./replays/action0x7a.w3g')
        expect(test.version).toBe('1.31')
        expect(test.players.length).toBe(1)
    })
})
