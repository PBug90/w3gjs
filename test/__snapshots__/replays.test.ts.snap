// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Replay parsing tests ignores a player entry in reforged extraPlayerList that misses in playerList 1`] = `
Array [
  Player {
    "_lastActionWasDeselect": false,
    "_lastRetrainingTime": 0,
    "_retrainingMetadata": Object {},
    "actions": Object {
      "ability": 54,
      "assigngroup": 87,
      "basic": 0,
      "buildtrain": 406,
      "esc": 1,
      "item": 0,
      "removeunit": 0,
      "rightclick": 3509,
      "select": 17,
      "selecthotkey": 1,
      "subgroup": 0,
      "timed": Array [
        46,
        248,
        260,
        210,
        253,
        249,
        249,
        257,
        261,
        117,
        258,
        267,
        255,
        125,
        208,
        251,
        252,
        235,
        74,
      ],
    },
    "apm": 222,
    "buildings": Object {
      "order": Array [],
      "summary": Object {},
    },
    "color": "#ff0303",
    "currentTimePlayed": 1100717,
    "heroCollector": Object {},
    "heroCount": 0,
    "heroes": Array [],
    "id": 1,
    "items": Object {
      "order": Array [],
      "summary": Object {},
    },
    "name": "Tallah#2192",
    "race": "H",
    "raceDetected": "H",
    "teamid": 0,
    "units": Object {
      "order": Array [],
      "summary": Object {},
    },
    "upgrades": Object {
      "order": Array [],
      "summary": Object {},
    },
  },
  Player {
    "_lastActionWasDeselect": false,
    "_lastRetrainingTime": 0,
    "_retrainingMetadata": Object {},
    "actions": Object {
      "ability": 21,
      "assigngroup": 0,
      "basic": 0,
      "buildtrain": 69,
      "esc": 0,
      "item": 0,
      "removeunit": 0,
      "rightclick": 1240,
      "select": 11,
      "selecthotkey": 0,
      "subgroup": 0,
      "timed": Array [
        55,
        64,
        82,
        54,
        25,
        58,
        94,
        128,
        76,
        43,
        30,
        188,
        31,
        167,
        101,
        16,
        79,
        37,
        13,
      ],
    },
    "apm": 74,
    "buildings": Object {
      "order": Array [],
      "summary": Object {},
    },
    "color": "#0042ff",
    "currentTimePlayed": 1092528,
    "heroCollector": Object {},
    "heroCount": 0,
    "heroes": Array [],
    "id": 2,
    "items": Object {
      "order": Array [],
      "summary": Object {},
    },
    "name": "RaNgeR#2873",
    "race": "H",
    "raceDetected": "H",
    "teamid": 0,
    "units": Object {
      "order": Array [],
      "summary": Object {},
    },
    "upgrades": Object {
      "order": Array [],
      "summary": Object {},
    },
  },
  Player {
    "_lastActionWasDeselect": false,
    "_lastRetrainingTime": 0,
    "_retrainingMetadata": Object {},
    "actions": Object {
      "ability": 74,
      "assigngroup": 94,
      "basic": 0,
      "buildtrain": 552,
      "esc": 0,
      "item": 0,
      "removeunit": 0,
      "rightclick": 4598,
      "select": 38,
      "selecthotkey": 0,
      "subgroup": 0,
      "timed": Array [
        128,
        325,
        334,
        342,
        367,
        342,
        338,
        326,
        345,
        172,
        329,
        311,
        325,
        182,
        268,
        284,
        270,
        288,
        80,
      ],
    },
    "apm": 290,
    "buildings": Object {
      "order": Array [],
      "summary": Object {},
    },
    "color": "#1ce6b9",
    "currentTimePlayed": 1108860,
    "heroCollector": Object {},
    "heroCount": 0,
    "heroes": Array [],
    "id": 3,
    "items": Object {
      "order": Array [],
      "summary": Object {},
    },
    "name": "Banjo#12494",
    "race": "H",
    "raceDetected": "H",
    "teamid": 0,
    "units": Object {
      "order": Array [],
      "summary": Object {},
    },
    "upgrades": Object {
      "order": Array [],
      "summary": Object {},
    },
  },
  Player {
    "_lastActionWasDeselect": false,
    "_lastRetrainingTime": 0,
    "_retrainingMetadata": Object {},
    "actions": Object {
      "ability": 0,
      "assigngroup": 0,
      "basic": 0,
      "buildtrain": 2,
      "esc": 0,
      "item": 0,
      "removeunit": 0,
      "rightclick": 0,
      "select": 5,
      "selecthotkey": 0,
      "subgroup": 0,
      "timed": Array [
        0,
        4,
        0,
        1,
        0,
        0,
        2,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
    },
    "apm": 0,
    "buildings": Object {
      "order": Array [],
      "summary": Object {},
    },
    "color": "#fe8a0e",
    "currentTimePlayed": 1108860,
    "heroCollector": Object {},
    "heroCount": 0,
    "heroes": Array [],
    "id": 4,
    "items": Object {
      "order": Array [],
      "summary": Object {},
    },
    "name": "TrencH#11363",
    "race": "H",
    "raceDetected": "H",
    "teamid": 0,
    "units": Object {
      "order": Array [],
      "summary": Object {},
    },
    "upgrades": Object {
      "order": Array [],
      "summary": Object {},
    },
  },
  Player {
    "_lastActionWasDeselect": false,
    "_lastRetrainingTime": 0,
    "_retrainingMetadata": Object {},
    "actions": Object {
      "ability": 47,
      "assigngroup": 0,
      "basic": 0,
      "buildtrain": 265,
      "esc": 0,
      "item": 0,
      "removeunit": 0,
      "rightclick": 3758,
      "select": 11,
      "selecthotkey": 0,
      "subgroup": 0,
      "timed": Array [
        30,
        230,
        245,
        243,
        240,
        260,
        231,
        248,
        251,
        114,
        274,
        264,
        250,
        117,
        223,
        257,
        249,
        254,
        101,
      ],
    },
    "apm": 221,
    "buildings": Object {
      "order": Array [],
      "summary": Object {},
    },
    "color": "#fffc00",
    "currentTimePlayed": 1108907,
    "heroCollector": Object {},
    "heroCount": 0,
    "heroes": Array [],
    "id": 11,
    "items": Object {
      "order": Array [],
      "summary": Object {},
    },
    "name": "second536#1404",
    "race": "H",
    "raceDetected": "H",
    "teamid": 0,
    "units": Object {
      "order": Array [],
      "summary": Object {},
    },
    "upgrades": Object {
      "order": Array [],
      "summary": Object {},
    },
  },
  Player {
    "_lastActionWasDeselect": false,
    "_lastRetrainingTime": 0,
    "_retrainingMetadata": Object {},
    "actions": Object {
      "ability": 40,
      "assigngroup": 1535,
      "basic": 0,
      "buildtrain": 238,
      "esc": 0,
      "item": 0,
      "removeunit": 0,
      "rightclick": 4807,
      "select": 11,
      "selecthotkey": 0,
      "subgroup": 0,
      "timed": Array [
        86,
        336,
        376,
        370,
        329,
        363,
        347,
        334,
        353,
        191,
        423,
        441,
        409,
        336,
        411,
        428,
        488,
        461,
        149,
      ],
    },
    "apm": 359,
    "buildings": Object {
      "order": Array [],
      "summary": Object {},
    },
    "color": "#20c000",
    "currentTimePlayed": 1108860,
    "heroCollector": Object {},
    "heroCount": 0,
    "heroes": Array [],
    "id": 5,
    "items": Object {
      "order": Array [],
      "summary": Object {},
    },
    "name": "Gravedigger#11445",
    "race": "H",
    "raceDetected": "H",
    "teamid": 1,
    "units": Object {
      "order": Array [],
      "summary": Object {},
    },
    "upgrades": Object {
      "order": Array [],
      "summary": Object {},
    },
  },
  Player {
    "_lastActionWasDeselect": false,
    "_lastRetrainingTime": 0,
    "_retrainingMetadata": Object {},
    "actions": Object {
      "ability": 37,
      "assigngroup": 12,
      "basic": 0,
      "buildtrain": 685,
      "esc": 0,
      "item": 0,
      "removeunit": 0,
      "rightclick": 3307,
      "select": 36,
      "selecthotkey": 20,
      "subgroup": 0,
      "timed": Array [
        95,
        252,
        237,
        276,
        271,
        224,
        235,
        245,
        196,
        105,
        306,
        248,
        263,
        150,
        153,
        298,
        221,
        252,
        70,
      ],
    },
    "apm": 223,
    "buildings": Object {
      "order": Array [],
      "summary": Object {},
    },
    "color": "#4a2a04",
    "currentTimePlayed": 1104098,
    "heroCollector": Object {},
    "heroCount": 0,
    "heroes": Array [],
    "id": 6,
    "items": Object {
      "order": Array [],
      "summary": Object {},
    },
    "name": "Niko#25549",
    "race": "H",
    "raceDetected": "H",
    "teamid": 1,
    "units": Object {
      "order": Array [],
      "summary": Object {},
    },
    "upgrades": Object {
      "order": Array [],
      "summary": Object {},
    },
  },
  Player {
    "_lastActionWasDeselect": false,
    "_lastRetrainingTime": 0,
    "_retrainingMetadata": Object {},
    "actions": Object {
      "ability": 33,
      "assigngroup": 269,
      "basic": 0,
      "buildtrain": 856,
      "esc": 0,
      "item": 0,
      "removeunit": 0,
      "rightclick": 4064,
      "select": 22,
      "selecthotkey": 9,
      "subgroup": 0,
      "timed": Array [
        203,
        299,
        345,
        306,
        359,
        339,
        287,
        312,
        273,
        123,
        345,
        300,
        318,
        181,
        272,
        353,
        253,
        349,
        36,
      ],
    },
    "apm": 289,
    "buildings": Object {
      "order": Array [],
      "summary": Object {},
    },
    "color": "#959697",
    "currentTimePlayed": 1089940,
    "heroCollector": Object {},
    "heroCount": 0,
    "heroes": Array [],
    "id": 7,
    "items": Object {
      "order": Array [],
      "summary": Object {},
    },
    "name": "Enclop#21668",
    "race": "H",
    "raceDetected": "H",
    "teamid": 1,
    "units": Object {
      "order": Array [],
      "summary": Object {},
    },
    "upgrades": Object {
      "order": Array [],
      "summary": Object {},
    },
  },
  Player {
    "_lastActionWasDeselect": false,
    "_lastRetrainingTime": 0,
    "_retrainingMetadata": Object {},
    "actions": Object {
      "ability": 0,
      "assigngroup": 0,
      "basic": 0,
      "buildtrain": 2,
      "esc": 0,
      "item": 0,
      "removeunit": 0,
      "rightclick": 0,
      "select": 4,
      "selecthotkey": 0,
      "subgroup": 0,
      "timed": Array [
        4,
        0,
        1,
        1,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ],
    },
    "apm": 0,
    "buildings": Object {
      "order": Array [],
      "summary": Object {},
    },
    "color": "#e55bb0",
    "currentTimePlayed": 1099704,
    "heroCollector": Object {},
    "heroCount": 0,
    "heroes": Array [],
    "id": 8,
    "items": Object {
      "order": Array [],
      "summary": Object {},
    },
    "name": "Ziggs#11403",
    "race": "H",
    "raceDetected": "H",
    "teamid": 1,
    "units": Object {
      "order": Array [],
      "summary": Object {},
    },
    "upgrades": Object {
      "order": Array [],
      "summary": Object {},
    },
  },
  Player {
    "_lastActionWasDeselect": false,
    "_lastRetrainingTime": 0,
    "_retrainingMetadata": Object {},
    "actions": Object {
      "ability": 24,
      "assigngroup": 0,
      "basic": 0,
      "buildtrain": 111,
      "esc": 0,
      "item": 0,
      "removeunit": 0,
      "rightclick": 3090,
      "select": 13,
      "selecthotkey": 0,
      "subgroup": 0,
      "timed": Array [
        10,
        193,
        183,
        172,
        181,
        209,
        232,
        197,
        167,
        106,
        234,
        220,
        196,
        148,
        131,
        201,
        193,
        191,
        74,
      ],
    },
    "apm": 175,
    "buildings": Object {
      "order": Array [],
      "summary": Object {},
    },
    "color": "#106246",
    "currentTimePlayed": 1109001,
    "heroCollector": Object {},
    "heroCount": 0,
    "heroes": Array [],
    "id": 10,
    "items": Object {
      "order": Array [],
      "summary": Object {},
    },
    "name": "BunsInHotDog#1757",
    "race": "H",
    "raceDetected": "H",
    "teamid": 1,
    "units": Object {
      "order": Array [],
      "summary": Object {},
    },
    "upgrades": Object {
      "order": Array [],
      "summary": Object {},
    },
  },
]
`;
